### PR TITLE
doc/cephfs: expand the role of quiesce in consistent backups

### PR DESCRIPTION
**Pull request by leonid-s-usov**
Fixes: https://tracker.ceph.com/issues/65881

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by zdover23** [comment]
Some light rewording (in a future, editorial PR, after this PR has been merged) will improve the English.

**Comment by leonid-s-usov** [comment]
@zdover23 why not now? Please comment with suggestions, let's save another PR roundtrip?

File: doc/cephfs/fs-volumes.rst
Comment (zdover23) [inline doc/cephfs/fs-volumes.rst:796]: s/Distributed applications that run against different mount points of the same Ceph file system
are facing a challenge of consistent snapshots for backups and disaster recovery. The challenge is due to
the fact that CephFS snapshots dont provide strong consistency guarantees when there are multiple
clients performing writes. Even when the application is using file system flushes to synchronize checkpoints across its distributed components, there aren't guarantees that all acked writes will be part of a given snapshot./Because CephFS snapshots do not provide strong-consistency guarantees in cases involving writes performed by multiple clients, distributed applications running against different mount points of the same Ceph file system face a serious challenge. Even in a case where an application uses file system flushes to synchronize checkpoints across its distributed components, there is no guarantee that all acknowledged writes will be part of a given snapshot./

File: doc/cephfs/fs-volumes.rst
Comment (zdover23) [inline doc/cephfs/fs-volumes.rst:801]: s/working with/that work with/

Comment (zdover23) [inline doc/cephfs/fs-volumes.rst:801]: s/The feature allows to pause IO to a set of subvolumes/The feature makes it possible to pause IO to a set of subvolumes/

File: doc/cephfs/fs-volumes.rst
Comment (zdover23) [inline doc/cephfs/fs-volumes.rst:804]: s/By reaching such a pause across all clients we can guarantee that any persistent checkpoints reached by the application will be recoverable from the snapshot group made during the pause./Enforcing such a pause across all clients makes it possible to guarantee that any persistent checkpoints reached by the application will be recoverable from the snapshot group made during the pause./

Comment (zdover23) [inline doc/cephfs/fs-volumes.rst:804]: s/We refer to the pause as `quiesce`/This pause is called a `quiesce`./

A full-stop (a period) is needed at the end of this sentence. I wonder if this is meant to be in double-quotes ("quiesce") or if it's meant to be in backticks, like this (``as though it is part of a command or filename``). This is up to you. It's dealer's choice.

File: doc/cephfs/fs-volumes.rst
Comment (zdover23) [inline doc/cephfs/fs-volumes.rst:806]: s/now//

(delete "now"--This document is going to live long enough that soon enough, "now" won't refer to what it referred to today.)

File: doc/cephfs/fs-volumes.rst
Comment (zdover23) [inline doc/cephfs/fs-volumes.rst:815]: I don't have any corrective comment to make here. I just think that this is a good change, and you should feel good about yourself for making it. I wish everyone cut out unnecessary words like this.

File: doc/cephfs/fs-volumes.rst
Comment (zdover23) [inline doc/cephfs/fs-volumes.rst:824]: s/The group/This set/

In cases like this (where I am ignorant of the brand-new technology), I think it's good to stick with the noun that was originally introduced. I always wonder in contexts like this if "group" means the same thing as "set". I think that changing "The group" to "This set" (that is, changing the definite article "The" to the demonstrative pronoun "This" makes it crystal clear that we're using this new sentence to talk about the set of subvolumes that we mentioned in the previous sentence.

**Comment by leonid-s-usov** [comment]
> > @zdover23 why not now? Please comment with suggestions, let's save another PR roundtrip?
> 
> The main reason that I had preferred not to do it now was that it's nearly midnight here and I have to be at the airport in ten hours. 

I appreciate your effort, thank you!

> The secondary reason that I had preferred not to do it now is that I have always found that editorial passes that improve the elegance of the English but don't affect the technical material are more expeditiously done in separate PRs, where all of the grammatical issues can be dealt with in one fell swoop.

I won't object to another pass if we need one. I also accept the reasoning.

> But what the hell? Let's try it this way.

Great, I prefer to do as much as possible in every iteration, so let's go

File: doc/cephfs/fs-volumes.rst
Comment (leonid-s-usov) [reply doc/cephfs/fs-volumes.rst:801]: done

Comment (leonid-s-usov) [reply doc/cephfs/fs-volumes.rst:801]: done

**Review by zdover23 — APPROVED** [review]

File: doc/cephfs/fs-volumes.rst
Comment (leonid-s-usov) [reply doc/cephfs/fs-volumes.rst:804]: done

**Comment by zdover23** [comment]
I approved this PR. It's usually my custom to approve PRs before all of the suggested changes have been made (because it is deeply, deeply frustrating to incorporate suggested corrections and then not get a response from the person who suggested the corrections). This way, you can just merge the thing after the force-pushing and squashing and you don't have to wait for sign-off.

Is this feature new in Squid? (I ask because I need to know which branches this should be backported to, or, more to the point, which branches this should not be backported to.)

File: doc/cephfs/fs-volumes.rst
Comment (leonid-s-usov) [reply doc/cephfs/fs-volumes.rst:804]: here's how a single backtick is rendered:
image

and here's how a double backtick is rendered:
image

I chose the former to highlight terms throughout the documentation, and the latter to denote literals sent or received through the CLI. I probably failed to be 100% consistent with this, but I tried.

Comment (leonid-s-usov) [reply doc/cephfs/fs-volumes.rst:804]: done

File: doc/cephfs/fs-volumes.rst
Comment (leonid-s-usov) [reply doc/cephfs/fs-volumes.rst:806]: done

File: doc/cephfs/fs-volumes.rst
Comment (leonid-s-usov) [reply doc/cephfs/fs-volumes.rst:815]: done. (I felt good about myself)

File: doc/cephfs/fs-volumes.rst
Comment (leonid-s-usov) [reply doc/cephfs/fs-volumes.rst:824]: done

File: doc/cephfs/fs-volumes.rst
Comment (zdover23) [reply doc/cephfs/fs-volumes.rst:815]: I feel good about you feeling good.

It's good.

**Comment by leonid-s-usov** [comment]
> Is this expansion of the quiesce feature new in Squid? (I ask because I need to know which branches this should be backported to, or, more to the point, which branches this should not be backported to.)

Yes, this will be backported to `squid` (the Ceph tracker is configured accordingly). In addition, this will be cherry-picked to the 7.1 downstream.

File: doc/cephfs/fs-volumes.rst
Comment (leonid-s-usov) [reply doc/cephfs/fs-volumes.rst:796]: This is a tricky one. I took the liberty to edit your variant, please review the new version. It's funny that you edited out the very reason we opened this PR: an explicit reference to backups and disaster recovery.

I'm adding it back because I like to stick to the plan 😅

Also, I'm surprised you suggest beginning a section with `Because`. I'm not qualified to argue, but I would like to double-check, and I do it boldly by removing the word in my edit - subject to your review.

**Comment by leonid-s-usov** [comment]
> I need a little more precision

Squid and only Squid

File: doc/cephfs/fs-volumes.rst
Comment (zdover23) [reply doc/cephfs/fs-volumes.rst]: The problem here is that "are facing a challenge of consistent snapshots for backups and disaster recovery" just isn't meaningful.

Here are some things that might be intended, and that make more sense in English:

(1) Distributed applications that run against different mountpoints of the same Ceph file system have difficulty creating consistent snapshots for backups and for disaster recovery.

(2) When distributed applications run against different mountpoints of the same Ceph file system, those applications have trouble creating consistent snapshots for backups and for disaster recovery.

The next sentence also uses the word "challenge" in a way that distinguished-level English just doesn't permit. I might suggest this:

For:

> The challenge is due to the fact that CephFS snapshots dont provide strong consistency guarantees when there are multiple clients performing writes.

I suggest:

> This is because CephFS snapshots don't provide strong-consistency guarantees when multiple clients are performing writes.

If backups and disaster recovery are the real star of the show, then I might suggest putting them up front in the first sentence of the section:

> When preparing backups and preparing for disaster recovery, there are some important challenges to bear in mind if you use  distributed applications that run against different mount points of the same Ceph file system.

In this context, you can use the word "challenges" and it doesn't sound wrong.

**Comment by leonid-s-usov** [comment]
As usually happens when I begin editing text I had written earlier, I included some refactoring for the subjective better. I hope this doesn't through us further back than where we started, but I would like to ask you to re-review the new version of this section as a whole, rather than only reviewing the [bits changed](https://github.com/ceph/ceph/compare/aa6b0ef43e55230662c022d777eff5868a92f82f..34ec2a4c6a065028d593b1e959600a3ce87541a8)

**Review by zdover23 — APPROVED** [review]
LGTM.

The first two paragraphs are now clear.

**Comment by leonid-s-usov** [comment]
> @leonid-s-usov Is this ready to merge now?

I think so, yes. We'll have to wait for the required checks to pass though.
In the meantime, do you know if I should open a BZ to backport this to 7.1? Or is there one already?

**Comment by zdover23** [comment]
> > @leonid-s-usov Is this ready to merge now?
> 
> I think so, yes. We'll have to wait for the required checks to pass though. In the meantime, do you know if I should open a BZ to backport this to 7.1? Or is there one already?

I don't know anything about downstream stuff. I work purely in the upstream.

**Comment by leonid-s-usov** [comment]
Ah, no, I need another review from someone in the downstream.

**Comment by zdover23** [comment]
> Ah, no, I need another review from someone in the downstream.

Is that IBM or Red Hat, or is there a difference nowadays?

**Comment by leonid-s-usov** [comment]
> > Ah, no, I need another review from someone in the downstream.
> 
> Is that IBM or Red Hat, or is there a difference nowadays?

IBM. I made this change per their request to get it picked to the 7.1 release. Just need a green light so that there's a single commit to backport

**Review by batrick — APPROVED** [review]

**Comment by zdover23** [comment]
https://github.com/ceph/ceph/pull/57390 - Squid backport
No Reef backport
No Quincy backport

**Pull request merged**
